File: src/main/java/tasks/Tinkoff/TinkoffGeneration/GraphsBfs/taskK.py
import heapq
import math
from collections import deque

n, m = map(int, input().split())

cities = list(map(int, input().split()))
graph = [[] for _ in range(n)]

for _ in range(m):
    u, v, w = map(int, input().split())
    u -= 1
    v -= 1
    graph[u].append((v, w))
    graph[v].append((u, w))

res = math.inf
res_frm, res_to = 0, 0
for start in range(n):
    if cities[start] != 1:
        continue

    dists = [math.inf] * n
    dists[start] = 0

    heap = [(0, start)]
    while heap:
        _, node = heapq.heappop(heap)
        if cities[node] + cities[start] == 3:
            if res > dists[node]:
                res = dists[node]
                res_frm, res_to = start, node
            break

        for nxt, weight in graph[node]:
            nxt_dist = dists[node] + weight
            if dists[nxt] > nxt_dist:
                heapq.heappush(heap, (nxt_dist, nxt))
                dists[nxt] = nxt_dist

if res == math.inf:
    print(-1)
else:
    print(res_frm + 1, res_to + 1, res)

File: src/main/java/tasks/Tinkoff/TinkoffGenerationС/GraphsBfs/taskK.cpp
#include <iostream>
#include <vector>
#include <set>

int main() {
    int n, m;
    std::cin >> n >> m;
    std::vector<int> cities(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> cities[i];
    }
    std::vector<std::vector<std::pair<int, int>>> graph(n);
    for (int i = 0; i < m; ++i) {
        int u, v, w;
        std::cin >> u >> v >> w;
        --u;
        --v;
        graph[u].push_back({v, w});
        graph[v].push_back({u, w});
    }
    int res = -1;
    int res_frm = 0, res_to = 0;

    std::vector<int> dists(n, -1);
    for (int start = 0; start < n; ++start) {
        if (cities[start] != 1) {
            continue;
        }

        dists[start] = 0;

        std::set<std::pair<long long, int>> heap;
        heap.insert({0, start});

        while (!heap.empty()) {
            int node = heap.begin()->second;
            heap.erase(heap.begin());

            if (cities[node] + cities[start] == 3) {
                if (res == -1 || res > dists[node]) {
                    res = dists[node];
                    res_frm = start;
                    res_to = node;
                }
                break;
            }
            for (auto [nxt, weight] : graph[node]) {
                int nxt_dist = dists[node] + weight;
                if (dists[nxt] == -1 || dists[nxt] > nxt_dist) {
                    heap.erase({dists[nxt], nxt});
                    dists[nxt] = nxt_dist;
                    heap.insert({dists[nxt], nxt});
                }
            }
        }
    }
    if (res == -1) {
        std::cout << -1 << std::endl;
    } else {
        std::cout << res_frm + 1 << " " << res_to + 1 << " " << res << std::endl;
    }
    return 0;
}

